src/Http/BaseApplication.php

```diff
@@ -61,6 +78,75 @@ public function __construct($configDir, EventManagerInterface $eventManager = nu
      */
     abstract public function middleware($middleware);
 
+    /**
+     * {@inheritDoc}
+     */
+    public function pluginEvents()
```

Comment:
I think that this method should behave similar to the `pluginMiddleware` and `pluginRoutes` ie accept an object and return the modified one.

The risk of dropping changes if a new instance is returned by mistake is the same as with the middleware and routes.

Reply:
Immutable approach of always using return value to continue sounds indeed more consistent.

Reply:
Ok. I will get this updated. Do you also think all the new hooks should have typehints added where we have existing interfaces?

Reply:
I've added typehints for `EventManagerInterface` as it was the only existing interface.

Reply:
@markstory `pluginEvents()` hasn't been updated with type hints.

Reply:
> It also doesn't take parameters or return a value either though.

`pluginRoutes` and `pluginMiddleware` take an argument and return the modified object. I don't like that `pluginEvents` is different. Either we remove that from routes and middleware, or add to events.

> Perhaps the Server and CommandRunner should drop their event managers and instead proxy the application manager?
In this case the application would need to support events. This is optional now, as http and console application interfaces don't extend `EventDispatcherInterface`.

Reply:
> It also doesn't take parameters or return a value either though.

`pluginRoutes` and `pluginMiddleware` take an argument and return the modified object. I don't like that `pluginEvents` is different. Either we remove that from routes and middleware, or add to events.

Also having `pluginEvents` with no arguments assumes that application that implements `PluginApplicationInterface` must have it's own event manager. Executing that method on an "eventless" application instance would take no effect. The idea of application "hooks" in my opinion is to take an object, modify it and return the modified instance.

> Perhaps the Server and CommandRunner should drop their event managers and instead proxy the application manager?

In this case the application would always need to support events. This is optional now, as http and console application interfaces don't extend `EventDispatcherInterface`.

Reply:
I feel like I need to explain my point a bit more.

In "isolation" application does not dispatch events by default (it doesn't have to implement the `EventDispatcherInterface`), the Server and CommandRunner do. So in this scenario event hooks should be executed on Server's and CommandRunner's event manager. So both `events` and `pluginEvents` must take an argument in order to modify the server/runner's event manager. 
If application does implement `EventDispatcherInterface` it's event manager is injected into the server/runner in `setApp` setter. The process of applying hooks on an event manager is still the same, but the subject (event manager instance) is different now (app's event manager)

Reply:
I hear you on it feeling inconsistent, my concern is how much logic adding a parameter/return value forces onto the caller of `pluginEvents()`. For example the bootstrap method on both `Server` and `CommmandRunner` would need to look like:

```php
        $this->app->bootstrap();
        if ($this->app instanceof EventApplicationInterface) {
            $eventManager = $this->app->events($this->getEventManager());
            $this->setEventManager($eventManager);
        }
        if ($this->app instanceof PluginApplicationInterface) {
            $this->app->pluginBootstrap();
            if ($this->app instanceof EventApplicationInterface) {
                $events = $this->app->getEventManager();
                $events = $this->app->pluginEvents($events);

                $this->app->setEventManager($events);
                $this->setEventManager($events);
            }
        }
```

To me this puts a lot of onus on the caller and breaks the encapsulation that the interfaces should be trying to provide. Perhaps the `PluginApplicationInterface`should include/consume the `EventApplicationInterface` and `EventManagerInterface`? That would allow us to have fewer checks across the framework. It would prevent 'only' using events and not plugins, but I think that is a reasonable tradeoff to make for reduced complexity.

Joining the interfaces may also allow us to make proxying eventmanager methods simpler on `Server` and `CommandRunner`. The `getEventManager()` method could have an implementation that looks something like:

```php
public function getEventManager()
{
  if ($this->app instanceof PluginApplicationInterface) {
    return $this->app->getEventManager();
  }
  return EventManager::instance();
}
```